README.md
# ShogiKoma

An image generator for Shogi's pieces (it's "ko-ma" in Japanese) for Ruby by cairo.

It supports one or two characters.

![fu1](http://myokoym.net/public/shogi_koma-fu1.png)
![fu2](http://myokoym.net/public/shogi_koma-fu2.png)

## Requirements

* [Ruby](https://www.ruby-lang.org/)
* [rcairo](https://github.com/rcairo/rcairo)
* [cairo](http://cairographics.org/)

## Installation

Add this line to your application's Gemfile:

    gem "shogi_koma"

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install shogi_koma

## Usage

### Write to PNG

```ruby
require "shogi_koma"

painter = ShogiKoma::Painter.new
#painter.width = 200        # as default
#painter.height = 200       # as default
#painter.font = "xxx"       # default: "Hiragino Mincho" or "MS Mincho" or "Noto"
painter.write_to_png("歩", "shogi_koma-fu1.png")
painter.write_to_png("歩兵", "shogi_koma-fu2.png")
```

### Draw to Cairo::Context

```ruby
require "cairo"
require "shogi_koma"

width  = 200
height = 200

Cairo::ImageSurface.new(:argb32, width, height) do |surface|
  Cairo::Context.new(surface) do |context|
    context.scale(width, height)
    painter = ShogiKoma::Painter.new
    painter.draw(context, "歩")
  end
end
```

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


NEWS.md
# NEWS

## 0.0.6: 2023-12-29

### Changes

  * Improvements
    * Fix default font
    * test: Use Rake::TestTask

## 0.0.5: 2014-08-18

### Changes

  * Improvements
    * Add a parameter "border_scale".

  * Fixes
    * Removed needless require.

## 0.0.4: 2014-01-05

### Changes

  * Improvements
    * Guess a font from a part of font name.

## 0.0.3: 2013-12-14

Supported customizing colors!

### Changes

  * Improvements
    * Supported customizing colors (body, frame and text).
      [Patch by Seiichi Yonezawa]

### Thanks

  * Seiichi Yonezawa

## 0.0.2: 2013-12-08

ASCII characters support release!

### Changes

  * Improvements
    * Supported ASCII characters.

## 0.0.1: 2013-12-06

Initial release!


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "shogi_koma"

require "test-unit"


test/painter_test.rb
require "test_helper"
require "cairo"

class PainterTest < Test::Unit::TestCase
  def setup
    @painter = ShogiKoma::Painter.new
  end

  class SetFontTest < self
    def test_found
      part_of_name = "M"
      assert_match(/#{part_of_name}/, @painter.set_font(part_of_name))
    end

    def test_not_found
      part_of_name = "ABCDE12345"
      assert_nil(@painter.set_font(part_of_name))
    end
  end

  class DrawTest < self
    def test_one_character
      assert_nothing_raised_in_draw("A")
    end

    def test_two_characters
      assert_nothing_raised_in_draw("AB")
    end

    def test_three_characters
      assert_nothing_raised_in_draw("ABC")
    end

    def test_body_color_with_rgb
      @painter.set_body_rgb(0.2, 0.2, 0.2)
      assert_nothing_raised_in_draw("D")
    end

    def test_text_color_with_name
      @painter.set_text_color(:deep_pink)
      assert_nothing_raised_in_draw("E")
    end

    def test_text_color_with_hex
      @painter.set_text_color("#ACE")
      assert_nothing_raised_in_draw("G")
    end

    private
    def assert_nothing_raised_in_draw(text)
      width =  200
      height = 200
      assert_nothing_raised do
        Cairo::ImageSurface.new(:argb32, width, height) do |surface|
          Cairo::Context.new(surface) do |context|
            context.scale(width, height)
            @painter.draw(context, text)
          end
        end
      end
    end
  end
end


test/test-painter.rb
require "shogi_koma/painter"
require "cairo"

class PainterTest < Test::Unit::TestCase
  def setup
    @painter = ShogiKoma::Painter.new
  end

  class SetFontTest < self
    def test_found
      part_of_name = "M"
      assert_match(/#{part_of_name}/, @painter.set_font(part_of_name))
    end

    def test_not_found
      part_of_name = "ABCDE12345"
      assert_nil(@painter.set_font(part_of_name))
    end
  end

  class DrawTest < self
    def test_one_character
      assert_nothing_raised_in_draw("A")
    end

    def test_two_characters
      assert_nothing_raised_in_draw("AB")
    end

    def test_three_characters
      assert_nothing_raised_in_draw("ABC")
    end

    def test_body_color_with_rgb
      @painter.set_body_rgb(0.2, 0.2, 0.2)
      assert_nothing_raised_in_draw("D")
    end

    def test_text_color_with_name
      @painter.set_text_color(:deep_pink)
      assert_nothing_raised_in_draw("E")
    end

    def test_text_color_with_hex
      @painter.set_text_color("#ACE")
      assert_nothing_raised_in_draw("G")
    end

    private
    def assert_nothing_raised_in_draw(text)
      width =  200
      height = 200
      assert_nothing_raised do
        Cairo::ImageSurface.new(:argb32, width, height) do |surface|
          Cairo::Context.new(surface) do |context|
            context.scale(width, height)
            @painter.draw(context, text)
          end
        end
      end
    end
  end
end


lib/shogi_koma/font.rb
require "fontdock"

module ShogiKoma
  class Font
    attr_reader :name
    def initialize(name="IPAMincho")
      set_font(name)
    end

    def set_font(part_of_font_name)
      name = find_font(part_of_font_name)
      @name = name if name
    end

    def find_font(part_of_font_name)
      found = find_font_from_pattern(/\A#{part_of_font_name}\z/)
      return found if found

      found = find_font_from_pattern(/\A#{part_of_font_name}\z/i)
      return found if found

      found = find_font_from_pattern(/\A#{part_of_font_name}/)
      return found if found

      found = find_font_from_pattern(/\A#{part_of_font_name}/i)
      return found if found

      found = find_font_from_pattern(/#{part_of_font_name}/)
      return found if found

      find_font_from_pattern(/#{part_of_font_name}/i)
    end

    private
    def find_font_from_pattern(pattern)
      Fontdock::Local.names.find do |name|
        pattern =~ name
      end
    end
  end
end


lib/shogi_koma/painter.rb
require "cairo"
require "fontdock"

module ShogiKoma
  class Painter
    attr_accessor :width, :height, :font, :border_scale
    attr_reader :body_color, :frame_color, :text_color
    def initialize
      @width = 200
      @height = 200
      ["Hiragino Mincho", "MS Mincho", "Noto"].each do |font_name|
        @font = Fontdock::Local.find(font_name)
        break if @font
      end
      @border_scale = 1.0
      set_body_rgb(1, 0.8, 0.2)
      set_frame_color(:black)
      set_text_color(:black)
    end

    def set_font(part_of_font_name)
      @font = Fontdock::Local.find(part_of_font_name)
    end

    def set_body_color(color)
      @body_color = Cairo::Color.parse(color)
    end

    def set_body_rgba(r, g, b, a=1.0)
      @body_color = Cairo::Color.parse([:rgba, r, g, b, a])
    end
    alias :set_body_rgb :set_body_rgba

    def set_frame_color(color)
      @frame_color = Cairo::Color.parse(color)
    end

    def set_frame_rgba(r, g, b, a=1.0)
      @frame_color = Cairo::Color.parse([:rgba, r, g, b, a])
    end
    alias :set_frame_rgb :set_frame_rgba

    def set_text_color(color)
      @text_color = Cairo::Color.parse(color)
    end

    def set_text_rgba(r, g, b, a=1.0)
      @text_color = Cairo::Color.parse([:rgba, r, g, b, a])
    end
    alias :set_text_rgb :set_text_rgba

    def write_to_png(text, output_path)
      Cairo::ImageSurface.new(:argb32, @width, @height) do |surface|
        Cairo::Context.new(surface) do |context|
          context.scale(@width, @height)
          draw(context, text)
        end
        surface.write_to_png(output_path)
      end
    end

    def draw(context, text)
      draw_body(context)
      text = divide(text)
      __send__("draw_text#{text.length}", context, text)
    end

    def draw_body(context)
      border_width = 0.01 * @border_scale
      context.set_line_width(border_width)
      context.move_to(0.2, 0.2)
      context.line_to(0.5, 0.1)
      context.line_to(0.8, 0.2)
      context.line_to(0.9, 0.9)
      context.line_to(0.1, 0.9)
      context.close_path
      context.set_source_color(@body_color)
      context.fill_preserve
      context.set_source_color(@frame_color)
      context.stroke
    end

    def draw_text1(context, text)
      context.set_source_color(@text_color)
      text = text[0] if text.is_a?(Array)
      context.select_font_face(@font)
      context.font_size = 0.6
      context.move_to(0.2, 0.75)
      context.show_text(text)
    end

    def draw_text2(context, text)
      context.set_source_color(@text_color)
      context.select_font_face(@font)
      context.font_size = 0.4
      context.move_to(0.3, 0.49)
      context.show_text(text[0])
      context.move_to(0.3, 0.85)
      context.show_text(text[1])
    end

    private
    def divide(text)
      case text.length
      when 1
        text
      when 2
        if text.bytes.to_a.length == 2
          [text]
        else
          text
        end
      else
        if text[0].bytes.to_a.length == 1 && text[1].bytes.to_a.length == 1
          [text[0..1], text[2..-1]]
        else
          [text[0], text[1..-1]]
        end
      end
    end
  end
end
